driver/src/order_driver.rs

```
@@ -1,108 +1,170 @@
+
+impl<'a, D: DbInterface, C: SnappContract> OrderProcessor<'a, D, C> {
+    pub fn new(db: &'a D, contract: &'a C, price_finder: &'a mut PriceFinding) -> Self {
+        OrderProcessor {
+            auction_bids: HashMap::new(),
+            db,
+            contract,
+            price_finder,
+        }
+    }
+
+    pub fn run(&mut self) -> Result<ProcessResult, DriverError> {
+        let auction_slot = self.contract.get_current_auction_slot()?;
+
+        info!("Current top auction slot is {:?}", auction_slot);
+        let slot = find_first_unapplied_slot(auction_slot, &|i| {
+            self.contract.has_auction_slot_been_applied(i)
+        })?;
+        if slot <= auction_slot {
+            info!("Highest unprocessed auction slot is {:?}", slot);
+            match batch_processing_state(slot, self.contract, &|i| {
+                self.contract.creation_timestamp_for_auction_slot(i)
+            })? {
+                ProcessingState::TooEarly => {
+                    info!("Need to wait before processing auction slot {:?}", slot)
+                }
+                ProcessingState::AcceptsBids => {
+                    if !self.auction_bids.contains_key(&slot) {
+                        self.bid_for_auction(slot)?;
+                        return Ok(ProcessResult::AuctionBid(slot));
+                    }
+                    info!("Already bid for auction slot {:?}", slot);
+                }
+                ProcessingState::AcceptsSolution => {
+                    if let Some(bid) = self.auction_bids.get(&slot) {
+                        self.contract.apply_auction(
+                            slot,
+                            bid.previous_state,
+                            bid.new_state,
+                            bid.solution.bytes(),
+                        )?;
+                        return Ok(ProcessResult::AuctionApplied(slot));
+                    }
+                    info!("Cannot find saved bid for auction slot {:?}", slot);
```

Should we return an error here?